hypothesis_regex.py

```diff
@@ -1,102 +1,268 @@
+            self._categories |= UNICODE_WORD_CATEGORIES
+            self._whitelist_chars.add('_')
+            if HAS_WEIRD_WORD_CHARS and self._unicode:
+                for c in UNICODE_WEIRD_NONWORD_CHARS:
+                    self._blacklist_chars.add(c)
+        elif category == sre.CATEGORY_NOT_WORD:
+            self._categories |= UNICODE_CATEGORIES - UNICODE_WORD_CATEGORIES
+            self._blacklist_chars.add('_')
+            if HAS_WEIRD_WORD_CHARS and self._unicode:
+                for c in UNICODE_WEIRD_NONWORD_CHARS:
+                    self._whitelist_chars.add(c)
+
+    def add_chars(self, chars):
+        for c in chars:
+            if self._ignorecase:
+                self._whitelist_chars.add(c.lower())
+                self._whitelist_chars.add(c.upper())
+            else:
+                self._whitelist_chars.add(c)
 
 
 @hs.defines_strategy
 def regex(regex):
-    """Return strategy that generates strings that match given regex."""
-    if hasattr(regex, 'pattern'):
-        regex = regex.pattern
+    """Return strategy that generates strings that match given regex.
+
+    Regex can be either a string or compiled regex (through `re.compile()`).
+
+    You can use regex flags (such as `re.IGNORECASE`, `re.DOTALL` or `re.UNICODE`)
+    to control generation. Flags can be passed either in compiled regex (specify
+    flags in call to `re.compile()`) or inside pattern with (?iLmsux) group.
+
+    Some tricky regular expressions are partly supported or not supported at all.
+    "^" and "$" do not affect generation. Positive lookahead/lookbehind groups
+    are considered normal groups. Negative lookahead/lookbehind groups do not do
+    anything.
+    """
+    if not hasattr(regex, 'pattern'):
+        regex = re.compile(regex)
 
-    codes = sre.parse(regex)
+    pattern = regex.pattern
+    flags = regex.flags
 
-    return ReprWrapperStrategy(_strategy(codes, {}), 'regex(%s)' % repr(regex))
+    codes = sre.parse(pattern)
 
+    return _strategy(codes, Context(flags=flags))
 
-def _strategy(codes, cache):
+
+def _strategy(codes, context):
     if not isinstance(codes, tuple):
-        return hs.tuples(*[_strategy(x, cache) for x in codes]).map(''.join)
+        strategies = []
+
+        i = 0
+        while i < len(codes):
+            if codes[i][0] == sre.LITERAL and not (context.flags & re.IGNORECASE):
+                j = i + 1
+                while j < len(codes) and codes[j][0] == sre.LITERAL:
+                    j += 1
+
+                strategies.append(hs.just(
+                    u''.join([six.unichr(charcode) for (_, charcode) in codes[i:j]])
+                ))
+
+                i = j
+            else:
+                strategies.append(_strategy(codes[i], context))
+                i += 1
+
+        if len(strategies) == 0:
+            return hs.just('')
+        elif len(strategies) == 1:
+            return strategies[0]
+        else:
+            return hs.tuples(*strategies).map(u''.join)
```

Comment:
`hs.tuples(*strategies).map(u''.join)` works correctly for length-zero and length-one tuples - `''.join(()) == ''` and `''.join(('a',)) == 'a'` - but this *is* slightly more efficient and makes it easier to lift literals later.